### ISSUE TITLE
Support building other projects as part of test image

### ISSUE THREAD

**Issue by dmnks** [rpm-software-management/rpm#2877]
Sometimes what you want to do is also build a dependency of RPM locally, just like RPM itself, and run the test-suite with that. A prime example being the rpm-sequoia project which is currently using RPM 4.18 (i.e. the previous test-suite version using fakechroot) in their CI, see https://github.com/rpm-software-management/rpm-sequoia/issues/61 for details. Another example would be popt which we also maintain.

As to how this can be facilitated by the test-suite, I'm not sure yet, but it shouldn't be *that* hard. Basically, we want to allow injecting custom stuff into the test image.

**Comment by pmatilai** [comment]
Yep, both rpm-sequoia and popt are something we'd occasionally want to build from sources instead of packaged versions. So this is actually fairly important piece, even if we have managed without it so far.

**Comment by dmnks** [comment]
Thinking about this for a moment, one simple solution comes to mind:

Allow for using a custom base image (or Dockerfile) for `make check`.

This way, our test image would be built on top of whatever custom image you've chosen. The only requirement would be for that image to itself be based on the same distro, e.g. the latest Fedora. That means, the project in question (e.g. rpm-sequoia) would have to provide their own Dockerfile that creates a minimal image with the project built from source and installed. To make use of it locally (or in CI), you'd simply set a cmake option that points to the Dockerfile (or just the given Podman/Docker tag). From there, `make check` would work as normal.

Does this make sense?

**Comment by pmatilai** [comment]
Seems I dropped the ball here, sorry. I suppose a custom base image could be handy.
It doesn't scale though, eg in case we'd want to build rpm-sequoia, popt and .. say, elfutils from sources to test a new feature before it's packaged in Fedora.

**Comment by dmnks** [comment]
> I suppose a custom base image could be handy. It doesn't scale though, eg in case we'd want to build rpm-sequoia, popt and .. say, elfutils from sources to test a new feature before it's packaged in Fedora.

Imagine for a moment that the test-suite doesn't run in containers (e.g. it still uses fakechroot). How would you accomplish the above?

**Comment by pmatilai** [comment]
Have local recipes from building from those upstream sources, install in suitable prefix with LD_LIBRARY_PATH. In a word, painfully? :smile:

**Comment by dmnks** [comment]
Thanks for taking the bait, the purpose of the (rhetorical) question was to get us closer to finding a solution :smile: 

What changes with the new test-suite is that you would do those builds in a container instead of your host, one that's preferably based on the test image. Then, you would run the test-suite manually from that container or commit the container as the "new" test image and run it with `make check` from the host as usual.

**Comment by dmnks** [comment]
Thing is, you still need to manually bother with containers and images. What we need here is an easy way to integrate this so you don't need to remember "oh, I first need to run a container, then commit it, blah blah".

One way perhaps is to use Toolbx. We just need to integrate `make check` into it so that you can run it from a Toolbx container directly. I've attempted this with (the now closed) #2830 but I'm planning on submitting a new version of that PR, so that could help.
